docs/design/query-caching.md

```diff
@@ -0,0 +1,224 @@
+
+The caching service should be database and endpoint agnostic. DAB will instantiate a singleton caching service that can be used within dependent services. In order to reduce coupling between DAB and a particular caching service implementation, the most generic caching options are publicly exposed in the runtime configuration: cache enablement and cache entry ttl. Additional options specific to the caching service will be explicitly set internally by DAB.
+
+Intentionally surfacing limited configuration options to the developer enables us to __make the caching service function uniformaly and independent of local hosting and SWA hosting configuration.__ This is because the developer exposed configuration properties are applied per-request and unnecessary for configuring the cache service during application startup.
+
+To reduce the sprawl of cache implementation across the code base, we will introduce the caching layer within __a common step of query execution for both REST and GraphQL: the query engine.__ The query engine's responsibility is to build a database query and dispatch the query to the appropriate query executor aligned to the backend database type. 
+
+By keeping the caching layer out of the query executor layer, we eliminate dependencies between database specific query execution logic and caching. Additionally, we prevent adding even tighter coupling between the runtime configuration and query executor because the query executor will not need to reference the runtime config to determine whether caching a request is necessary.
+
+### Cache entry
+
+#### Key
+
+`dataSourceName:queryText:queryParameters`
+
+The following example illustrates a real value created using the key schema listed above (new lines added for readability):
+
+```text
+7a07f92a-1aa2-4e2a-81d6-b9af0a25bbb6:
+
+SELECT TOP 1 [dbo_books].[id] AS [id], [dbo_books].[title] AS [title], [dbo_books].[publisher_id] AS [publisher_id] FROM [dbo].[books] AS [dbo_books] WHERE [dbo_books].[id] = @param0 ORDER BY [dbo_books].[id] ASC FOR JSON PATH, INCLUDE_NULL_VALUES,WITHOUT_ARRAY_WRAPPER:
+
+{"@param0":{"Value":1,"DbType":11},"@param1":{"Value":"id","DbType":null},"@param2":{"Value":"title","DbType":null},"@param3":{"Value":"publisher_id","DbType":null}}
+```
+
+The goals of the created key:
+- Uniquely store query text per data source to avoid collisions.
+- Agnostic towards GraphQL or REST request metadata.
+- Determinate size based on the size of the string generated.
+- Equivalent query metadata will map to the same cache value.
+  - This protects against reference type comparison where two distinct C# objects with same property values (request metadata) are treated as two separate cache entries instead of the same entry.
+
+- Rely on the underlying cache implementation to store the key/value pair in a concurrency compatible collection. The implementation defined collection will abstract away the work done to handle cache key hash collisions and appropriate bucketing.
```

Comment:
It would be helpful to note that the implementation would make use of `FusionCache`